host/crypto/bcrypt/cert.c

```diff
@@ -0,0 +1,1295 @@
+                der_blob.pbData,
+                der_blob.cbData,
+                CERT_STORE_ADD_REPLACE_EXISTING,
+                NULL))
+        {
+            OE_RAISE_MSG(
+                OE_CRYPTO_ERROR,
+                "CertAddEncodedCertificateToStore failed, err=%#x\n",
+                GetLastError());
+        }
+
+        found_certs = true;
+
+        free(der_blob.pbData);
+        der_blob.pbData = NULL;
+        der_blob.cbData = 0;
+    }
+
+    if (!found_certs)
+        OE_RAISE(
+            OE_INVALID_PARAMETER,
+            "No certs could be loaded from provided PEM.\n",
+            NULL);
+
+    *cert_store = load_cert_store;
+    load_cert_store = NULL;
+    result = OE_OK;
+
+done:
+    if (load_cert_store)
+        CertCloseStore(load_cert_store, 0);
+
+    if (pem_cert)
+        free(pem_cert);
+
+    if (der_blob.pbData)
+    {
+        der_blob.pbData = NULL;
+        der_blob.cbData = 0;
+    }
+
+    return result;
+}
+
+/* Verify each certificate in the chain against its predecessor. */
+static oe_result_t _bcrypt_get_cert_chain(
+    PCCERT_CONTEXT cert_context,
+    HCERTSTORE trusted_store,
+    size_t expected_chain_length,
+    PCCERT_CHAIN_CONTEXT* cert_chain)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    PCCERT_CHAIN_CONTEXT found_chain = NULL;
+
+    /* CertGetCertificateChain can increment the refcount on cert_context
+     * by >1 due to CERT_CHAIN_CACHE_END_CERT. Do not force close the store,
+     * which will destroy the cached cert and corrupt future cert chaining
+     * calls.
+     */
+    if (!CertGetCertificateChain(
+            NULL, /* use the default engine */
+            cert_context,
+            NULL, /* use the default time */
+            trusted_store,
+            (PCERT_CHAIN_PARA)&_OE_DEFAULT_CERT_CHAIN_PARAMS,
+            _OE_DEFAULT_CERT_CHAIN_FLAGS,
+            NULL, /* reserved */
+            &found_chain))
+    {
+        DWORD err = GetLastError();
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR, "CertGetCertificateChain failed, err=%#x\n", err);
+    }
+
+    if (found_chain && found_chain->cChain > 0 &&
+        (expected_chain_length == _OE_CERT_CHAIN_LENGTH_ANY ||
+         found_chain->rgpChain[0]->cElement == expected_chain_length))
+    {
+        *cert_chain = found_chain;
+        found_chain = NULL;
+        result = OE_OK;
+    }
+    else
+        result = OE_NOT_FOUND;
+
+done:
+    if (found_chain)
+        CertFreeCertificateChain(found_chain);
+
+    return result;
+}
+
+static oe_result_t _bcrypt_check_revocation(
+    PCCERT_CONTEXT cert,
+    PCERT_CHAIN_ELEMENT* cert_issuers,
+    DWORD cert_issuers_count,
+    HCERTSTORE cert_store)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    bool is_validated = false;
+    DWORD flags = _OE_DEFAULT_GET_CRL_FLAGS;
+    PCCRL_CONTEXT prev_crl = NULL;
+    PCCRL_CONTEXT crl = CertGetCRLFromStore(
+        cert_store, cert_issuers[0]->pCertContext, prev_crl, &flags);
```

Comment:
Can we add a check for crl == null here. I think it works correctly in your impl, since you have a while and then an if. So this is just a suggestion

Reply:
I'm not sure what the suggestion is ... is it to encapsulate the `while` loop in the `if (crl)` block? If so, that's not the intention, because the `while` loop can take non-null `crl` variable and set it to null if it cannot find a base CRL, so there still needs to be an `if (crl)` check after the `while` loop.